rosco-manifests/src/main/java/com/netflix/spinnaker/rosco/manifests/TemplateUtils.java

```diff
@@ -27,56 +27,75 @@
-    } catch (NoSuchAlgorithmException e) {
-      throw new RuntimeException("Failed to save bake manifest: " + e.getMessage(), e);
-    }
-  }
+    private RetrySupport retrySupport = new RetrySupport();
 
-  protected Path downloadArtifactToTmpFile(BakeManifestEnvironment env, Artifact artifact) throws IOException {
-    if (artifact.getReference() == null) {
-      throw new InvalidRequestException("Input artifact has an empty 'reference' field.");
+    private String nameFromReference(String reference) {
+        try {
+            MessageDigest md = MessageDigest.getInstance("MD5");
+            return DatatypeConverter.printHexBinary(md.digest(reference.getBytes()));
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException("Failed to save bake manifest: " + e.getMessage(), e);
+        }
     }
-    Path path = Paths.get(env.getStagingPath().toString(), nameFromReference(artifact.getReference()));
-    OutputStream outputStream = new FileOutputStream(path.toString());
 
-    Response response = retrySupport.retry(() -> clouddriverService.fetchArtifact(artifact), 5, 1000, true);
+    protected Path downloadArtifactToTmpFile(BakeManifestEnvironment env, Artifact artifact) throws IOException {
+        if (artifact.getReference() == null) {
+            throw new InvalidRequestException("Input artifact has an empty 'reference' field.");
+        }
+        Path path = Paths.get(env.getStagingPath().toString(), nameFromReference(artifact.getReference()));
+        OutputStream outputStream = new FileOutputStream(path.toString());
+        Response response = retrySupport.retry(() -> clouddriverService.fetchArtifact(artifact), 5, 1000, true);
+        if (response.getBody() != null) {
+            InputStream inputStream = response.getBody().in();
+            IOUtils.copy(inputStream, outputStream);
+            inputStream.close();
+        }
+        outputStream.close();
 
-    if (response.getBody() != null) {
-      InputStream inputStream = response.getBody().in();
-      IOUtils.copy(inputStream, outputStream);
-      inputStream.close();
+        return path;
     }
-    outputStream.close();
 
-    return path;
-  }
+    protected void downloadArtifactToTmpFileStructure(BakeManifestEnvironment env, Artifact artifact) throws IOException {
+        if (artifact.getReference() == null) {
+            throw new InvalidRequestException("Input artifact has an empty 'reference' field.");
+        }
+        Path artifactPath = Paths.get(artifact.getReference());
+        String filename = artifactPath.getFileName().toString();
+        String subfolder = artifactPath.toString().replace(filename, "");
+        Path tmpPath = Paths.get(env.getStagingPath().resolve(subfolder).toString());
+        Files.createDirectories(tmpPath);
+        File newfile = new File(env.getStagingPath().resolve(subfolder).resolve(filename).toString());
+        newfile.createNewFile();
```

Comment:
I don't think that's what you want, e.g. if `artifactPath = "test/test"` -> after replace `"/"`. If you just want the parent:
```
Path tmpPath = Paths.get(env.getStagingPath().resolve(artifactPath.getParent()).toString()
File newfile = new File(tmpPath.resolve(artifactPath.getFileName().toString());
```
(untested!)

Reply:
replaced.